File: 543-diameter-of-binary-tree/543-diameter-of-binary-tree.py
class Solution:
    def diameterOfBinaryTree(self, root: TreeNode) -> int:
        res=0
        def height(root):
            nonlocal res
            if not root:
                return 0
            
            left = height(root.left)
            right = height(root.right)
            res = max(res, left + right)
            return 1+max(left, right)
        
        height(root)
        return res

File: 543-diameter-of-binary-tree/543-diameter-of-binary-tree.java
public class Solution {
    int res= 0;
    
    public int diameterOfBinaryTree(TreeNode root) {
        help(root);
        return res;
    }
    
    private int help(TreeNode root) {
        if (root == null) return 0;
        
        int left =  help(root.left);
        int right = help(root.right);
        
        res = Math.max(res, left + right);
        
        return 1+ Math.max(left, right);
    }
}